### PR TITLE
Use DB to support ordered channel, ICA and packet fee

### PR DESCRIPTION
**Pull request by ALPAC-4**
Use sqlite to store indexed events to support:
- Ordered Channel
- Channel Creation (to support ICA)
- Packet Fee

### CODE REVIEW

File: src/lib/config.ts [hunk 1]
```diff
@@ -10,7 +10,52 @@ export interface Config {
...
 
-export const syncInfoFile = env.SYNC_INFO || './.syncInfo'
+interface ChainConfig {
+  bech32Prefix: string
+  chainId: string
+  gasPrice: string
+  lcdUri: string
+  rpcUri: string
+  wallets: WalletConfig[]
+  feeFilter?: PacketFee
```
Comment (beer-1) [inline src/lib/config.ts:23]: Fee filter is chain config, but the example config has it in wallet config level

File: src/db/controller/channel.ts [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+export class ChannelController {
+  static tableName = 'channel_open_close'
+  public static async feedEvents(
+    lcd: LCDClient,
+    chainId: string,
+    events: ChannelOpenCloseEvent[]
+  ): Promise<() => void> {
+    const feedFns: (() => void)[] = []
+    for (const event of events) {
```
Comment (beer-1) [inline src/db/controller/channel.ts:22]: If the purpose of this function is handling promises in concurrently, may better to use
```suggestion
    const feedFns: Promise<(() => void)>[] = []

    for (const event of events) {
      switch (event.type) {
        case 'channel_open_init':
          feedFns.push(this.feedChannelOpenInitEvent(lcd, chainId, event))
          break
          ...
```

File: src/db/controller/channel.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    return () => {
+      del<ChannelOpenCloseTable>(DB, this.tableName, [
+        {
+          state: ChannelState.CLOSE,
+          chain_id: chainId,
+          port_id: event.channelOpenCloseInfo.dstPortId,
+          channel_id: event.channelOpenCloseInfo.dstChannelId,
+        },
+      ])
+      PacketController.updateTimeout(
```
Comment (beer-1) [inline src/db/controller/channel.ts:306]:
```suggestion
      // Mark all packets as timed out
      PacketController.updateTimeout(
```

File: src/db/controller/channel.ts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    event: ChannelOpenCloseEvent
+  ): Promise<() => void> {
+    const connection = await ConnectionController.getConnection(
+      lcd,
+      chainId,
+      event.channelOpenCloseInfo.srcConnectionId
+    )
+
+    // add channel on open for dst chain
+    const channelOnOpen: ChannelOpenCloseTable = {
```
Comment (beer-1) [inline src/db/controller/channel.ts:273]: `channelOnOpenClose` to avoid confusion

File: src/db/controller/channel.ts [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+      port_id: event.channelOpenCloseInfo.srcPortId,
+      channel_id: event.channelOpenCloseInfo.srcChannelId,
+      counterparty_chain_id: chainId,
+      counterparty_connection_id: event.channelOpenCloseInfo.dstConnectionId,
+      counterparty_port_id: event.channelOpenCloseInfo.dstPortId,
+      counterparty_channel_id: event.channelOpenCloseInfo.dstChannelId,
+    }
+
+    return () => {
+      del<ChannelOpenCloseTable>(DB, this.tableName, [
```
Comment (beer-1) [inline src/db/controller/channel.ts:185]: no case for deleting wrong entry? don't we need to check `chain_id`?

File: src/db/controller/channel.ts [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+      counterparty_chain_id: chainId,
+      counterparty_connection_id: event.channelOpenCloseInfo.srcConnectionId,
+      counterparty_port_id: event.channelOpenCloseInfo.srcPortId,
+      counterparty_channel_id: event.channelOpenCloseInfo.srcChannelId,
+    }
+
+    return () => {
+      del<ChannelOpenCloseTable>(DB, this.tableName, [
+        {
+          state: ChannelState.INIT,
```
Comment (beer-1) [inline src/db/controller/channel.ts:226]: maybe need to delete TRY_OPEN?

File: src/db/controller/channel.ts (hunk 2 above)
Comment (ALPAC-4) [reply src/db/controller/channel.ts:22]: The purpose of this function is that remove async function to execute all SQL in a single db transaction.

File: src/db/controller/channel.ts (hunk 5 above)
Comment (ALPAC-4) [reply src/db/controller/channel.ts:185]: (`state`, `counterparty_chain_id`, `counterparty_channel_id`) is unique

File: src/db/controller/channel.ts [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+      counterparty_port_id: event.channelOpenCloseInfo.srcPortId,
+      counterparty_channel_id: event.channelOpenCloseInfo.srcChannelId,
+    }
+
+    return () => {
+      insert(DB, this.tableName, channelOnOpen)
+    }
+  }
+
+  private static async feedChannelCloseConfirmEvent(
```
Comment (beer-1) [inline src/db/controller/channel.ts:305]: seems unnecessary `async`

File: src/db/controller/channel.ts (hunk 5 above)
Comment (beer-1) [reply src/db/controller/channel.ts:185]: oh reasonable

File: src/db/controller/channel.ts (hunk 2 above)
Comment (beer-1) [reply src/db/controller/channel.ts:22]: okay if that is the case, good to use db transaction.

File: src/db/controller/channel.ts [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+    if (events.length === 0) return
+
+    del<ChannelOpenCloseTable>(
+      DB,
+      this.tableName,
+      events.map((v) => ({ id: v.id as number }))
+    )
+  }
+
+  public static updateInProgress(id?: number, inProgress = true) {
```
Comment (beer-1) [inline src/db/controller/channel.ts:113]: id can be undefined?

File: src/db/controller/channel.ts [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    chainId: string,
+    event: ChannelOpenCloseEvent
+  ): Promise<() => void> {
+    const connection = await ConnectionController.getConnection(
+      rest,
+      chainId,
+      event.channelOpenCloseInfo.dstConnectionId
+    )
+    return () => {
+      del<ChannelOpenCloseTable>(DB, this.tableName, [
```
Comment (beer-1) [inline src/db/controller/channel.ts:257]: hmm is this for cleanup?? 

in the flow, I thought each channel delete an event created by counterparty chain.

File: src/db/controller/client.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    // update client
+    client.revision_height = parseInt(
+      event.consensusHeights.split(',')[0].split('-')[1]
+    )
+
+    if (header.signedHeader?.header?.time.seconds) {
+      client.last_update_time = Number(header.signedHeader.header.time.seconds)
+    }
+
+    insert(DB, this.tableName, client)
```
Comment (beer-1) [inline src/db/controller/client.ts:62]: update?

File: src/workers/wallet.ts [hunk 11]
```diff
@@ -0,0 +1,672 @@
...
+        accountNumber: this.accountNumber,
+      })
+
+      const result = await this.wallet.rest.tx.broadcast(signedTx)
+
+      if (isTxError(result)) {
+        if (result.raw_log.startsWith('account sequence mismatch')) {
+          try {
+            const expected = result.raw_log.split(', ')[1]
+            this.sequence = Number(expected.split(' ')[1])
```
Comment (beer-1) [inline src/workers/wallet.ts:324]: This will be affected at L343.

File: src/workers/wallet.ts [hunk 12]
```diff
@@ -0,0 +1,672 @@
...
+
+  private async filterWriteAckPackets(
+    writeAckPackets: PacketWriteAckTable[]
+  ): Promise<PacketWriteAckTable[]> {
+    // create path => packet map
+    const writeAckPacketMap: Record<string, PacketWriteAckTable[]> = {}
+    const writeAckPacketsToDel: PacketWriteAckTable[] = []
+
+    for (const packet of writeAckPackets) {
+      const path = `${packet.src_port}/${packet.src_port}`
```
Comment (beer-1) [inline src/workers/wallet.ts:486]: maybe `src_channel_id/src_port`?

File: src/workers/wallet.ts [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+
+  private async filterTimeoutPackets(
+    timeoutPackets: PacketTimeoutTable[] | PacketSendTable[]
+  ): Promise<PacketTimeoutTable[]> {
+    // create path => packet map
+    const timeoutPacketMap: Record<string, PacketTimeoutTable[]> = {}
+    const timeoutPacketsToDel: PacketTimeoutTable[] = []
+
+    for (const packet of timeoutPackets) {
+      const path = `${packet.src_port}/${packet.src_port}`
```
Comment (beer-1) [inline src/workers/wallet.ts:534]: `src_channel_id/src_port`

**Review by beer-1 — COMMENTED** [review]
overall works are LGTM, just left minor comments and will create separate PR for typo and some small changes.

File: src/db/controller/packet.ts [hunk 14]
```diff
@@ -0,0 +1,622 @@
...
+        {
+          dst_chain_id: connection.counterparty_chain_id,
+          dst_connection_id: connection.counterparty_connection_id,
+          dst_channel_id: event.packetInfo.dstChannel,
+          sequence: Number(event.packetInfo.sequence),
+        },
+      ])
+
+      // remove packet timeout
+      del<PacketTimeoutTable>(DB, this.tableNamePacketSend, [
```
[the remark this answers is not in the document]
Comment (ALPAC-4) [reply src/db/controller/packet.ts:570]: I think that case is impossible cause timeout proof will not return if already `recvPacket` done.

**Review by beer-1 — APPROVED** [review]
LGTM

**Pull request merged**